package/samplers/auto_sampler/sampler.py

```diff
@@ -0,0 +1,220 @@
+class _ThreadLocalSampler(threading.local):
+    def __init__(self, sampler: BaseSampler) -> None:
+        self._sampler = sampler
+        super().__init__()
+
+
+class AutoSampler(BaseSampler):
+    _N_COMPLETE_TRIALS_FOR_CMAES = 250
+
+    """Sampler automatically choosing an appropriate sampler based on search space.
+
+    This sampler is convenient when you are unsure what sampler to use.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+            import optunahub
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -5, 5)
+                return x**2
+
+            module = optunahub.load_module("samplers/auto_sampler")
+            study = optuna.create_study(sampler=module.AutoSampler())
+            study.optimize(objective, n_trials=300)
+
+    .. note::
+        This sampler requires optional dependencies of Optuna.
+        You can install them with ``pip install optunahub scipy torch cmaes``.
+        Alternatively, you can install them with ``pip install -r https://hub.optuna.org/samplers/auto_sampler/requirements.txt``.
+
+    Args:
+        seed: Seed for random number generator.
+        constraints_func:
+            An optional function that computes the objective constraints. It must take a
+            :class:`~optuna.trial.FrozenTrial` and return the constraints. The return value must
+            be a sequence of :obj:`float` s. A value strictly larger than 0 means that a
+            constraints is violated. A value equal to or smaller than 0 is considered feasible.
+            If ``constraints_func`` returns more than one value for a trial, that trial is
+            considered feasible if and only if all values are equal to 0 or smaller.
+
+            The ``constraints_func`` will be evaluated after each successful trial.
+            The function won't be called when trials fail or they are pruned, but this behavior is
+            subject to change in the future releases.
+    """
+
+    def __init__(
+        self,
+        *,
+        seed: int | None = None,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+    ) -> None:
+        self._rng = LazyRandomState(seed)
+        seed_for_random_sampler = self._rng.rng.randint(MAXINT32)
+        sampler: BaseSampler = RandomSampler(seed=seed_for_random_sampler)
+        self._thread_local_sampler = _ThreadLocalSampler(sampler)
+        self._constraints_func = constraints_func
+
+    @property
+    def _sampler(self) -> BaseSampler:
+        return self._thread_local_sampler._sampler
+
+    @_sampler.setter
+    def _sampler(self, sampler: BaseSampler) -> None:
+        self._thread_local_sampler._sampler = sampler
+
+    def reseed_rng(self) -> None:
+        self._rng.rng.seed()
+        self._sampler.reseed_rng()
+
+    def _include_conditional_param(self, study: Study) -> bool:
+        trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE, TrialState.PRUNED))
```

Comment:
How about using `_get_trials` with `use_cache=True`?

```suggestion
        trials = study._get_trials(deepcopy=False, states=(TrialState.COMPLETE, TrialState.PRUNED))
```